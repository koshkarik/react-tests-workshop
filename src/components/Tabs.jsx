import React from 'react';
import _ from 'lodash';
import uuid from 'uuid';
import cookies from 'js-cookie';
import Modal from 'react-responsive-modal';
import RssParser from 'rss-parser';
import 'bootstrap/dist/css/bootstrap.css';
// import 'react-lumberjack';
import {
  Tab,
  Tabs,
  TabList,
  TabPanel,
} from 'react-tabs';
import 'react-tabs/style/react-tabs.css';

const parser = new RssParser();

const proxyCorsServer = 'https://cors-anywhere.herokuapp.com/';

const modalStyle = {
  modal: {
    minWidth: '400px !important',
  },
};

export default class MyTabs extends React.Component {
  constructor(props) {
    super(props);
    const activeTabIndex = Number(cookies.get('activeTabIndex')) || 0;

    this.state = {
      activeTabIndex,
      rssModalOpen: false,
      isRssFetching: false,
      error: null,
      rssUrl: 'http://feeds.bbci.co.uk/news/world/rss.xml',
      tabs: [
        {
          title: 'Tab 1',
          description: 'content 1',
          id: uuid(),
        },
        {
          title: 'Tab 2',
          description: 'content 2',
          id: uuid(),
        },
      ],
    };
  }

  getRss = async (url) => {
    const rssUrl = `${proxyCorsServer}${url}`;
    const feed = await parser.parseURL(rssUrl);
    return feed.items.map(item => item.title);
  }

  handleRssUrlChange = (event) => {
    const { value } = event.target;

    this.setState({
      rssUrl: value,
    });
  }

  onRemove = id => () => {
    this.setState(prevState => ({
      tabs: _.reject(prevState.tabs, { id }),
    }));
  }

  onRssSubmit = (e) => {
    const { rssUrl } = this.state;

    e.preventDefault();
    this.setState({
      isRssFetching: true,
      error: null,
    }, async () => {
      try {
        const rss = await this.getRss(rssUrl);
        const newTab = {
          title: rssUrl,
          description: rss.join('\n'),
          id: uuid(),
        };
        this.setState(prevState => ({
          tabs: [...prevState.tabs, newTab],
        }));
        this.onCloseRssModal();
      } catch (err) {
        this.setState({
          error: 'Failed to load rss from that url',
          isRssFetching: false,
        });
      }
    });
  }

  onRemoveError = () => {
    this.setState({
      error: null,
    });
  }

  onCreateTab = () => {
    const newTab = {
      title: `title ${Math.random()}`,
      description: `description ${Math.random()}`,
      id: uuid(),
    };

    this.setState(prevState => ({
      tabs: [...prevState.tabs, newTab],
    }));
  }

  onSelectTab = (tabIndex) => {
    this.setState({
      activeTabIndex: tabIndex,
    });
    cookies.set('activeTabIndex', tabIndex);
  }


  onOpenRssModal = () => {
    this.setState({
      rssModalOpen: true,
    });
  }

  onCloseRssModal = () => {
    this.setState({
      rssModalOpen: false,
      isRssFetching: false,
      rssUrl: '',
      error: null,
    });
  }

  renderRemoveButton = id => (
    <button type="button" className="btn btn-outline-dark ml-2" onClick={this.onRemove(id)}>X</button>
  )

  renderTabs = () => {
    const { tabs } = this.state;
    return tabs.map(({ id, title }) => (
      <Tab key={id} data-test="tab-control">
        {title}
        {this.renderRemoveButton(id)}
      </Tab>
    ));
  }

  renderTabPanels = () => {
    const { tabs } = this.state;
    return tabs.map(({ id, description }) => (
      <TabPanel key={id} data-test="tab-content">
        {description}
      </TabPanel>
    ));
  }

  renderError = () => {
    const { error } = this.state;
    return (
      <div className="alert alert-danger" role="alert">
        {error}
        <button type="button" className="close" data-dismiss="alert" aria-label="Close" onClick={this.onRemoveError}>
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
    );
  }

  renderRssModal = () => {
    const { rssModalOpen, error } = this.state;
    if (!rssModalOpen) {
      return null;
    }

    return (
      <Modal open styles={modalStyle} onClose={this.onCloseRssModal} center>
        <h4>Add tab from Rss</h4>
        {this.renderRssForm()}
        {error && this.renderError()}
      </Modal>
    );
  }

  renderSpinner = () => (
    <div className="spinner-border ml-2" role="status">
      <span className="sr-only">Loading...</span>
    </div>
  );

  renderRssForm = () => {
    const { isRssFetching, rssUrl } = this.state;
    return (
      <form onSubmit={this.onRssSubmit}>
        <div className="input-group mt-4 mb-4">
          <input
            type="text"
            className="form-control"
            value={rssUrl}
            onChange={this.handleRssUrlChange}
            placeholder="Enter rss url"
            aria-label="Recipient's username"
            aria-describedby="button-addon2"
          />
          <div className="input-group-append">
            <button
              disabled={isRssFetching}
              className="btn btn-secondary"
              type="submit"
              id="button-addon2"
            >
              {'Add'}
            </button>
            {isRssFetching && this.renderSpinner()}
          </div>
        </div>
      </form>
    );
  }

  render() {
    const { activeTabIndex } = this.state;
    return (
      <div className="mt-5">
        <button
          data-test="add-tab-btn"
          className="btn btn-primary"
          type="button"
          onClick={this.onCreateTab}
        >
          {'Add new fake Tab'}
        </button>
        <button
          data-test="add-tab-btn"
          className="btn btn-success ml-4"
          type="button"
          onClick={this.onOpenRssModal}
        >
          {'Add new RSS tab'}
        </button>
        <div className="mt-5">
          <Tabs data-test="tabs-container" selectedIndex={activeTabIndex} onSelect={this.onSelectTab}>
            <TabList>
              {this.renderTabs()}
            </TabList>
            {this.renderTabPanels()}
          </Tabs>
        </div>
        {this.renderRssModal()}
      </div>
    );
  }
}
