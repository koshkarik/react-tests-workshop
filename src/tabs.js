export default [
  {
    key: 0,
    label: 'Batman',
    content: 'At vero eos et accusamus et iusto odio dignissimos ducimus',
  },
  {
    key: 1,
    label: 'Superman',
    content: 'Et harum quidem rerum facilis est et expedita distinctio.',
  },
  {
    key: 2,
    label: 'Ironman',
    content: 'Nam libero tempore, cum soluta nobis est eligendi optio, cumque nihil impedit',
  },
];
