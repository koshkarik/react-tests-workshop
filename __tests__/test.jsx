import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Tabs from '../src/components/Tabs';
import tabs from '../src/tabs';

Enzyme.configure({ adapter: new Adapter() });

test('example', () => {
  const wrapper = mount(<Tabs tabs={tabs} />);
  expect(wrapper.render()).toMatchSnapshot();

  const tab1 = wrapper.find('[data-testid="tab_0"] > li');
  const tab2 = wrapper.find('[data-testid="tab_1"] > li');
  const tab3 = wrapper.find('[data-testid="tab_2"] > li');

  console.log(tab1.debug({ ignoreProps: true }));

  tab1.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  tab2.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  tab3.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
});
