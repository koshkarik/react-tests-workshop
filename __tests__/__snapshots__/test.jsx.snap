// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example 1`] = `
<div
  class="react-tabs"
  data-tabs="true"
>
  <ul
    class="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="true"
      class="react-tabs__tab react-tabs__tab--selected"
      data-testid="tab_0"
      id="react-tabs-0"
      role="tab"
      tabindex="0"
    >
      Batman
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="false"
      class="react-tabs__tab"
      data-testid="tab_1"
      id="react-tabs-2"
      role="tab"
    >
      Superman
    </li>
    <li
      aria-controls="react-tabs-5"
      aria-disabled="false"
      aria-selected="false"
      class="react-tabs__tab"
      data-testid="tab_2"
      id="react-tabs-4"
      role="tab"
    >
      Ironman
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-0"
    class="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="react-tabs-1"
    role="tabpanel"
  >
    At vero eos et accusamus et iusto odio dignissimos ducimus
  </div>
  <div
    aria-labelledby="react-tabs-2"
    class="react-tabs__tab-panel"
    id="react-tabs-3"
    role="tabpanel"
  />
  <div
    aria-labelledby="react-tabs-4"
    class="react-tabs__tab-panel"
    id="react-tabs-5"
    role="tabpanel"
  />
</div>
`;

exports[`example 2`] = `
<div
  class="react-tabs"
  data-tabs="true"
>
  <ul
    class="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="true"
      class="react-tabs__tab react-tabs__tab--selected"
      data-testid="tab_0"
      id="react-tabs-0"
      role="tab"
      tabindex="0"
    >
      Batman
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="false"
      class="react-tabs__tab"
      data-testid="tab_1"
      id="react-tabs-2"
      role="tab"
    >
      Superman
    </li>
    <li
      aria-controls="react-tabs-5"
      aria-disabled="false"
      aria-selected="false"
      class="react-tabs__tab"
      data-testid="tab_2"
      id="react-tabs-4"
      role="tab"
    >
      Ironman
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-0"
    class="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="react-tabs-1"
    role="tabpanel"
  >
    At vero eos et accusamus et iusto odio dignissimos ducimus
  </div>
  <div
    aria-labelledby="react-tabs-2"
    class="react-tabs__tab-panel"
    id="react-tabs-3"
    role="tabpanel"
  />
  <div
    aria-labelledby="react-tabs-4"
    class="react-tabs__tab-panel"
    id="react-tabs-5"
    role="tabpanel"
  />
</div>
`;

exports[`example 3`] = `
<div
  class="react-tabs"
  data-tabs="true"
>
  <ul
    class="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="false"
      class="react-tabs__tab"
      data-testid="tab_0"
      id="react-tabs-0"
      role="tab"
    >
      Batman
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="true"
      class="react-tabs__tab react-tabs__tab--selected"
      data-testid="tab_1"
      id="react-tabs-2"
      role="tab"
      tabindex="0"
    >
      Superman
    </li>
    <li
      aria-controls="react-tabs-5"
      aria-disabled="false"
      aria-selected="false"
      class="react-tabs__tab"
      data-testid="tab_2"
      id="react-tabs-4"
      role="tab"
    >
      Ironman
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-0"
    class="react-tabs__tab-panel"
    id="react-tabs-1"
    role="tabpanel"
  />
  <div
    aria-labelledby="react-tabs-2"
    class="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="react-tabs-3"
    role="tabpanel"
  >
    Et harum quidem rerum facilis est et expedita distinctio.
  </div>
  <div
    aria-labelledby="react-tabs-4"
    class="react-tabs__tab-panel"
    id="react-tabs-5"
    role="tabpanel"
  />
</div>
`;

exports[`example 4`] = `
<div
  class="react-tabs"
  data-tabs="true"
>
  <ul
    class="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="false"
      class="react-tabs__tab"
      data-testid="tab_0"
      id="react-tabs-0"
      role="tab"
    >
      Batman
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="false"
      class="react-tabs__tab"
      data-testid="tab_1"
      id="react-tabs-2"
      role="tab"
    >
      Superman
    </li>
    <li
      aria-controls="react-tabs-5"
      aria-disabled="false"
      aria-selected="true"
      class="react-tabs__tab react-tabs__tab--selected"
      data-testid="tab_2"
      id="react-tabs-4"
      role="tab"
      tabindex="0"
    >
      Ironman
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-0"
    class="react-tabs__tab-panel"
    id="react-tabs-1"
    role="tabpanel"
  />
  <div
    aria-labelledby="react-tabs-2"
    class="react-tabs__tab-panel"
    id="react-tabs-3"
    role="tabpanel"
  />
  <div
    aria-labelledby="react-tabs-4"
    class="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="react-tabs-5"
    role="tabpanel"
  >
    Nam libero tempore, cum soluta nobis est eligendi optio, cumque nihil impedit
  </div>
</div>
`;
